### ISSUE TITLE
Deserialization behavior change with Records, `@JsonCreator` and `@JsonValue` between 2.17 and 2.18

### ISSUE THREAD

**Issue by arlampin** [FasterXML/jackson-databind#4724]
### Search before asking

- [X] I searched in the [issues](https://github.com/FasterXML/jackson-databind/issues) and found nothing similar.

### Describe the bug

Using both JsonCreator and JsonValue annotations with Java Records makes ObjectMapper seemingly skip single-arg static JsonCreator method and try to contract instance directly. I couldn't find any hint that this was a planned change.

### Version Information

2.18.0

### Reproduction

The following code works with 2.17.2, but fails with 2.18.0

```java
    @Test
    void deserialization() throws Exception {
        new ObjectMapper().readValue("\"\"", Something.class);
    }

    public record Something(String value) {

        public Something {
            if (value == null || value.isEmpty()) {
                throw new IllegalArgumentException("Value cannot be null or empty");
            }
        }

        @JsonCreator
        public static Something of(String value) {
            if (value.isEmpty()) {
                return null;
            }
            return new Something(value);
        }

        @Override
        @JsonValue
        public String toString() {
            return value;
        }

    }

``` 

Removing JsonValue annotation makes deserialization to work again.

**Comment by ianfp** [comment]
This also happens in Kotlin. It seems to have something to do with data classes that have secondary constructors. This test reproduces it:

```kotlin
import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertDoesNotThrow
import java.util.UUID
import java.util.UUID.randomUUID

data class MyThing(val id: UUID, val status: ThingStatus, val child: ChildThing) {
    constructor(status: ThingStatus, child: ChildThing) : this(randomUUID(), status, child)
}

enum class ThingStatus {
    COOL,
    NOT_COOL,
}

data class ChildThing(val id: UUID)

class MyThingTest {
    private val json = jacksonObjectMapper()
        .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)

    @Test
    fun fromJson() {
        assertDoesNotThrow {
            json.readValue<MyThing>(
                """
            {
                "id": "${randomUUID()}",
                "status": "NOT_COOL",
                "child": {
                    "id": "${randomUUID()}"
                }
            }
                """.trimIndent(),
            )
        }
    }
}
```

That test passes on 2.17; fails on 2.18.

**Comment by cowtowncoder** [comment]
That looks like a bug for sure: I would expect:

1. Deserialization to use `@JsonCreator` annotated factory method
2. Serialization to use value returned by `@JsonValue` annotated method.

One to note tho: it may (altho should not!) be necessary to use

    @JsonCreator(mode = JsonCreator.Mode.DELEGATING)

if the idea is to deserialize from JSON String (instead of single-property JSON Object).

**Comment by cowtowncoder** [comment]
One quick thing to check: would behavior be different if `@JsonValue` annotated was anything other than `toString()`? Just trying to rule out possibilities of what may be going on.

I also assume this same case would work for regular POJOs.

**Comment by cowtowncoder** [comment]
@ianfp Let's not assume it's same thing, your case looks very different from OP's.
In fact not 100% sure why you think it is same thing.

You may want to file separate issue against jackson-module-kotlin wrt your issue.

**Comment by arlampin** [comment]
1. `mode = JsonCreator.Mode.DELEGATING` doesn't seem to have any effect
2. The name of the method doesn't seem to have any effect either
3. Regular POJO _does_ work

**Comment by cowtowncoder** [comment]
Thank you for checking @arlampin! These are about as I expected -- but wanted to rule out some possibilities.

Failing test added via #4726, to hopefully help fix the issue.

**Comment by yihtserns** [comment]
What we know:
1. When the constructor has only 1-arg + there is `@JsonValue` annotation, currently that constructor becomes a delegating creator: https://github.com/FasterXML/jackson-databind/blob/26bdfec3dd07c809e910d6360a6bf3ea764cc286/src/test-jdk17/java/com/fasterxml/jackson/databind/records/RecordImplicitCreatorsTest.java#L236-L261
2. In previous versions, a constructor/method annotated with `@JsonCreator` will have precedence, which is why `Something.of` is used as the delegating creator - seems like this precedence is likely swapped in `2.18.0`.

While #​2 is out of their control, a (desperate) user can do something with #​1 to work around the regression:
```java
public record Something(...) {

    @JsonCreator(mode = Mode.DISABLED)
    public Something {
        ...
    }
    ...
}
```

**Comment by JooHyukKim** [comment]
> In previous versions, a constructor/method annotated with @JsonCreator will have precedence, which is why Something.of is used as the delegating creator - seems like this precedence is likely swapped in 2.18.0.

Do you think there is chance that solution is as simple as swapping back the precedence, @yihtserns? 
I am afraid not, but just in case 😀.

**Comment by cowtowncoder** [comment]
Changing precedence not intentional; but this wasn't a case covered by unit tests it seems.

Records have "default" or "preferred" constructor, but that should have lower precedence than explicitly annotated constructor or factory method.

**Comment by JooHyukKim** [comment]
Fix against 2.18? @cowtowncoder

**Comment by cowtowncoder** [comment]
Yes, unless drastic changes were needed, fix for 2.18.1 it all possible.
